src/Libraries/Microsoft.Extensions.AI.Abstractions/Utilities/AIJsonUtilities.Schema.Transform.cs

```diff
@@ -0,0 +1,190 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#pragma warning disable LA0001 // Use the 'Microsoft.Shared.Diagnostics.Throws' class instead of explicitly throwing exception for improved performance
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text.Json;
+using System.Text.Json.Nodes;
+using Microsoft.Shared.Diagnostics;
+
+namespace Microsoft.Extensions.AI;
+
+public static partial class AIJsonUtilities
+{
+    /// <summary>
+    /// Transforms the given JSON schema based on the provided options.
+    /// </summary>
+    /// <param name="schema">The schema document to transform.</param>
+    /// <param name="transformOptions">The options governing schema transformation.</param>
+    /// <returns>A new schema document with transformations applied.</returns>
+    /// <remarks>The schema and any nested schemas are transformed using depth-first traversal.</remarks>
+    public static JsonElement TransformSchema(JsonElement schema, AIJsonSchemaTransformOptions transformOptions)
+    {
+        _ = Throw.IfNull(transformOptions);
+
+        if (transformOptions == AIJsonSchemaTransformOptions.Default)
+        {
+            throw new ArgumentException("The options instance does not specify any transformations.", nameof(transformOptions));
+        }
+
+        JsonNode? nodeSchema = JsonSerializer.SerializeToNode(schema, JsonContext.Default.JsonElement);
+        List<string>? path = transformOptions.TransformSchemaNode is not null ? [] : null;
+        JsonNode transformedSchema = TransformSchemaCore(nodeSchema, transformOptions, path);
+        return JsonSerializer.Deserialize(transformedSchema, JsonContext.Default.JsonElement);
+    }
+
+    private static JsonNode TransformSchemaCore(JsonNode? schema, AIJsonSchemaTransformOptions transformOptions, List<string>? path)
+    {
+        switch (schema?.GetValueKind())
+        {
+            case JsonValueKind.False:
+                if (transformOptions.ConvertBooleanSchemas)
+                {
+                    schema = new JsonObject { [NotPropertyName] = (JsonNode)true };
```

Comment:
Is this cast from bool to JsonNode allocating?

Reply:
Yes, it allocates a new `JsonNode` instance every time. We pretty much need to do this because `JsonNode` stores a reference to the parent node that needs to be unique.